File: app/src/main/java/com/jbrunton/mymovies/app/moviedetails/MovieDetailsViewState.java
package com.jbrunton.mymovies.app.moviedetails;

import com.google.auto.value.AutoValue;
import com.jbrunton.mymovies.app.movies.MovieViewState;
import com.jbrunton.mymovies.app.shared.LoadingViewState;

@AutoValue
public abstract class MovieDetailsViewState {
    public abstract LoadingViewState loadingViewState();
    public abstract MovieViewState movie();

    public static MovieDetailsViewState.Builder builder() {
        return new AutoValue_MovieDetailsViewState.Builder();
    }

    public static MovieDetailsViewState buildLoadingState() {
        return builder()
                .setLoadingViewState(LoadingViewState.LOADING_STATE)
                .setMovie(MovieViewState.EMPTY)
                .build();
    }

    @AutoValue.Builder
    public abstract static class Builder {
        public abstract Builder setLoadingViewState(LoadingViewState loadingViewState);
        public abstract Builder setMovie(MovieViewState movie);
        public abstract MovieDetailsViewState build();
    }
}

File: app/src/main/java/com/jbrunton/mymovies/moviedetails/MovieDetailsViewState.kt
package com.jbrunton.mymovies.moviedetails

import com.jbrunton.entities.Movie
import com.jbrunton.mymovies.movies.MovieViewState
import com.jbrunton.mymovies.movies.from
import com.jbrunton.mymovies.shared.LoadingViewState

typealias MovieDetailsViewState = LoadingViewState<MovieViewState>

fun Movie.toViewState(): MovieViewState {
    val overview = this.overview.get()
    return MovieViewState.Builder().from(this).apply {
        this.overview = overview
    }.build()
}